[UnifiedPDF] Introduce AsyncPDFRenderer

smfr opened the pull request:
#### 98c763e246170ff480978642d91f27ff6bff79e8
<pre>
[UnifiedPDF] Introduce AsyncPDFRenderer
<a href="https://bugs.webkit.org/show_bug.cgi?id=269368">https://bugs.webkit.org/show_bug.cgi?id=269368</a>
<a href="https://rdar.apple.com/122950491">rdar://122950491</a>

Reviewed by Tim Horton.

This is the first step to rendering PDF content off the main thread.

AsyncPDFRenderer is a TiledBackingClient, so gets notified when the TileBacking needs
to repaint a tile. This causes it to enqueue a request to render a tile (identified by
a grid index/tile index pair in `TileForGrid`) on a workqueue (currently serial, but likely
concurrent in future). The tile rendering request includes a `PDFPageCoverage`, which
describes the set of PDF pages and their geometry which will go into this tile.

Rendering a tile on the work queue involves allocating an in-process ImageBuffer, calling
PDFDocument for the PDFPages for this tile, and drawing them (without selection). That
ImageBuffer is then transferred back to the main thread, and put into a cache, again
keyed by `TileForGrid`.

When a tile render is complete, we need to trigger a repaint so we can paint the tile
in the normal painting code path. There&apos;s a little trickiness to avoid continual repaints,
which is handled by `AsyncPDFRenderer::willRepaintTile()`.

* Source/WebKit/Platform/Logging.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::pdfDocument const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.h: Added.
(WTF::TileForGridHash::hash):
(WTF::TileForGridHash::equal):
(WTF::HashTraits&lt;WebKit::TileForGrid&gt;::emptyValue):
(WTF::HashTraits&lt;WebKit::TileForGrid&gt;::deletedValue):
(WTF::HashTraits&lt;WebKit::TileForGrid&gt;::constructDeletedValue):
(WTF::HashTraits&lt;WebKit::TileForGrid&gt;::isDeletedValue):
(WebKit::AsyncPDFRenderer::paintingWorkQueue const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm: Added.
(WebKit::AsyncPDFRenderer::create):
(WebKit::AsyncPDFRenderer::AsyncPDFRenderer):
(WebKit::AsyncPDFRenderer::~AsyncPDFRenderer):
(WebKit::AsyncPDFRenderer::setupWithLayer):
(WebKit::AsyncPDFRenderer::willRepaintTile):
(WebKit::AsyncPDFRenderer::willRemoveTile):
(WebKit::AsyncPDFRenderer::willRepaintAllTiles):
(WebKit::AsyncPDFRenderer::clearCachedTiles):
(WebKit::AsyncPDFRenderer::enqueuePaintWithClip):
(WebKit::AsyncPDFRenderer::paintTileOnWorkQueue):
(WebKit::AsyncPDFRenderer::paintPDFIntoBuffer):
(WebKit::AsyncPDFRenderer::transferBufferToMainThread):
(WebKit::AsyncPDFRenderer::paintTileForClip):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:

Canonical link: <a href="https://commits.webkit.org/274659@main">https://commits.webkit.org/274659@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/212eba5a140bdfd81954536dca4334c23ca1f78c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39673 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18652 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42029 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42208 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35573 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21600 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15981 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33114 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40247 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/15774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34330 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13642 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43485 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/36064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35617 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39402 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14485 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11930 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16091 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8890 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16140 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15748 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 reviewed (approved):
A bit messy but I know more patches are coming.

smfr committed:
[UnifiedPDF] Introduce AsyncPDFRenderer
https://bugs.webkit.org/show_bug.cgi?id=269368
rdar://122950491

Reviewed by Tim Horton.

This is the first step to rendering PDF content off the main thread.

AsyncPDFRenderer is a TiledBackingClient, so gets notified when the TileBacking needs
to repaint a tile. This causes it to enqueue a request to render a tile (identified by
a grid index/tile index pair in `TileForGrid`) on a workqueue (currently serial, but likely
concurrent in future). The tile rendering request includes a `PDFPageCoverage`, which
describes the set of PDF pages and their geometry which will go into this tile.

Rendering a tile on the work queue involves allocating an in-process ImageBuffer, calling
PDFDocument for the PDFPages for this tile, and drawing them (without selection). That
ImageBuffer is then transferred back to the main thread, and put into a cache, again
keyed by `TileForGrid`.

When a tile render is complete, we need to trigger a repaint so we can paint the tile
in the normal painting code path. There's a little trickiness to avoid continual repaints,
which is handled by `AsyncPDFRenderer::willRepaintTile()`.

* Source/WebKit/Platform/Logging.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::pdfDocument const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.h: Added.
(WTF::TileForGridHash::hash):
(WTF::TileForGridHash::equal):
(WTF::HashTraits<WebKit::TileForGrid>::emptyValue):
(WTF::HashTraits<WebKit::TileForGrid>::deletedValue):
(WTF::HashTraits<WebKit::TileForGrid>::constructDeletedValue):
(WTF::HashTraits<WebKit::TileForGrid>::isDeletedValue):
(WebKit::AsyncPDFRenderer::paintingWorkQueue const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm: Added.
(WebKit::AsyncPDFRenderer::create):
(WebKit::AsyncPDFRenderer::AsyncPDFRenderer):
(WebKit::AsyncPDFRenderer::~AsyncPDFRenderer):
(WebKit::AsyncPDFRenderer::setupWithLayer):
(WebKit::AsyncPDFRenderer::willRepaintTile):
(WebKit::AsyncPDFRenderer::willRemoveTile):
(WebKit::AsyncPDFRenderer::willRepaintAllTiles):
(WebKit::AsyncPDFRenderer::clearCachedTiles):
(WebKit::AsyncPDFRenderer::enqueuePaintWithClip):
(WebKit::AsyncPDFRenderer::paintTileOnWorkQueue):
(WebKit::AsyncPDFRenderer::paintPDFIntoBuffer):
(WebKit::AsyncPDFRenderer::transferBufferToMainThread):
(WebKit::AsyncPDFRenderer::paintTileForClip):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:

Canonical link: https://commits.webkit.org/274659@main
